cache/redis_client.go

```diff
@@ -0,0 +1,463 @@
+		c.duration.WithLabelValues(opSet).Observe(time.Since(start).Seconds())
+	})
+
+	if errors.Is(err, errMemcachedAsyncBufferFull) {
+		c.skipped.WithLabelValues(opSet, reasonAsyncBufferFull).Inc()
+		level.Debug(c.logger).Log("msg", "failed to store item to redis because the async buffer is full", "err", err, "size", len(c.asyncQueue))
+		return nil
+	}
+	return err
+}
+
+// GetMulti implement RemoteCacheClient.
+func (c *redisClient) GetMulti(ctx context.Context, keys []string) map[string][]byte {
+	if len(keys) == 0 {
+		return nil
+	}
+	start := time.Now()
+	results := make(map[string][]byte, len(keys))
+	var mu sync.Mutex
+	var total int
+
+	err := doWithBatch(ctx, len(keys), c.config.GetMultiBatchSize, c.getMultiGate, func(startIndex, endIndex int) error {
+		currentKeys := keys[startIndex:endIndex]
+		resp, err := c.MGet(ctx, currentKeys...).Result()
+		if err != nil {
+			level.Warn(c.logger).Log("msg", "failed to mget items from redis", "err", err, "items", len(resp))
+			return nil
+		}
+		mu.Lock()
+		defer mu.Unlock()
+		for i := 0; i < len(resp); i++ {
+			key := currentKeys[i]
+			switch val := resp[i].(type) {
+			case string:
+				total += len(val)
+				results[key] = stringToBytes(val)
+			case nil: // miss
+			default:
+				level.Warn(c.logger).Log("msg",
+					fmt.Sprintf("unexpected redis mget result type:%T %v", resp[i], resp[i]))
+			}
+		}
+		return nil
+	})
+	if err != nil {
+		level.Warn(c.logger).Log("msg", "failed to mget items from redis", "err", err, "items", len(keys))
+		return nil
+	}
+	c.dataSize.WithLabelValues(opGetMulti).Observe(float64(total))
+	c.duration.WithLabelValues(opGetMulti).Observe(time.Since(start).Seconds())
+	return results
+}
+
+// Delete implement RemoteCacheClient.
+func (c *redisClient) Delete(ctx context.Context, key string) error {
+	errCh := make(chan error, 1)
+
+	enqueueErr := c.enqueueAsync(func() {
+		start := time.Now()
+		c.operations.WithLabelValues(opDelete).Inc()
+
+		err := c.Del(ctx, key).Err()
+		if err != nil {
+			level.Debug(c.logger).Log(
+				"msg", "failed to delete redis item",
+				"key", key,
+				"err", err,
+			)
+			c.trackError(opDelete, err)
+		} else {
+			c.duration.WithLabelValues(opDelete).Observe(time.Since(start).Seconds())
+		}
+		errCh <- err
+	})
+
+	if errors.Is(enqueueErr, errMemcachedAsyncBufferFull) {
+		c.skipped.WithLabelValues(opDelete, reasonAsyncBufferFull).Inc()
+		level.Debug(c.logger).Log("msg", "failed to delete redis item because the async buffer is full", "err", enqueueErr, "size", len(c.asyncQueue))
+		return enqueueErr
+	}
+	// Wait for the delete operation to complete.
+	return <-errCh
+}
+
+// Stop implement RemoteCacheClient.
+func (c *redisClient) Stop() {
+	close(c.stop)
+
+	// Wait for all workers to finish.
+	c.workers.Wait()
+
+	if err := c.Close(); err != nil {
+		level.Error(c.logger).Log("msg", "redis close err")
+	}
+}
+
+func (c *redisClient) enqueueAsync(op func()) error {
```

Comment:
I think all the logic around the async queue is the same with memcached. Can be moved to a dedicated struct (e.g. `clientAsyncQueue` and generalised there.